### ISSUE TITLE
Multiple running database instances on same database file

### ISSUE THREAD

**Issue by hannesmuehleisen** [cwida/duckdb#77]
It should be possible for multiple processes to open and change the same database directory.

**Comment by hannesmuehleisen** [comment]
Restrict to multiple read-only

**Comment by hannesmuehleisen** [comment]
#40 related

**Comment by hannesmuehleisen** [comment]
https://www.sqlite.org/faq.html#q5 states

> Multiple processes can have the same database open at the same time. Multiple processes can be doing a SELECT at the same time. But only one process can be making changes to the database at any moment in time, however.

> SQLite uses reader/writer locks to control access to the database. (Under Win95/98/ME which lacks support for reader/writer locks, a probabilistic simulation is used instead.) But use caution: this locking mechanism might not work correctly if the database file is kept on an NFS filesystem. This is because fcntl() file locking is broken on many NFS implementations. You should avoid putting SQLite database files on NFS if multiple processes might try to access the file at the same time. On Windows, Microsoft's documentation says that locking may not work under FAT filesystems if you are not running the Share.exe daemon. People who have a lot of experience with Windows tell me that file locking of network files is very buggy and is not dependable. If what they say is true, sharing an SQLite database between two or more Windows machines might cause unexpected problems.

> We are aware of no other embedded SQL database engine that supports as much concurrency as SQLite. SQLite allows multiple processes to have the database file open at once, and for multiple processes to read the database at once. When any process wants to write, it must lock the entire database file for the duration of its update. But that normally only takes a few milliseconds. Other processes just wait on the writer to finish then continue about their business. Other embedded SQL database engines typically only allow a single process to connect to the database at once.

> However, client/server database engines (such as PostgreSQL, MySQL, or Oracle) usually support a higher level of concurrency and allow multiple processes to be writing to the same database at the same time. This is possible in a client/server database because there is always a single well-controlled server process available to coordinate access. If your application has a need for a lot of concurrency, then you should consider using a client/server database. But experience suggests that most applications need much less concurrency than their designers imagine.

> When SQLite tries to access a file that is locked by another process, the default behavior is to return SQLITE_BUSY. You can adjust this behavior from C code using the sqlite3_busy_handler() or sqlite3_busy_timeout() API functions.

**Comment by hannesmuehleisen** [comment]
We should use fcntl() and windows equivalent to make sure there are only readers or only one writer. Benefit of fcntl() is that it cleans up the locks when process dies.

**Comment by Mytherin** [comment]
Implemented by e402ad80313cc284de81712e1757a9e01c6fa601, multiple readers or single writer of the same database file are allowed

**Comment by Mytherin** [comment]
> Concurrent connections from multiple processes in Windows on the same .db file is still not working in version 0.2.7 I needed read only access to same db file while another process was writing to it. This functionality still fails for me. I found a work around by hacking the function below this whole morning and modifying it so I can get the Windows build to allow multiple processes as readers with one process writing. Please consider including this fix, which will add a lot of value to Windows users of DuckDB. Concurrent access is immensely usefull for a lot of scenarios and I am surprised this was not tested on Windows. Presented below is the changed function that will allow Windows version to work concurrently, please consider it to be included in next release as it will fix the Windows concurrency issue. The problem is the shared mode flags combination you used in your code, Windows only allows certain combinations as valid and yours was not one of them, unfortunatelly for me. Since I spent all this time chasing the cause of this problem I thought you might benefit from my findings. Thanks for all your efforts creating DuckDB, this is an impressive product and immensely usefull for a lot of scenarios.

As mentioned in this thread, multiple readers OR a single writer of the same database file are allowed right now. Multiple readers AND a single writer is not. The original code is correct. The changes you propose here appears to change the behavior of the locks so that multiple readers and writers do not lock each-other out.

While that might work as expected in many situations since the writer will append to the WAL most of the time, having both readers and writers active at the same time could potentially cause odd behavior when the writer performs a checkpoint of the database while the readers are in the middle of reading from the database file. This is particularly problematic when you have a long-running reader process with a database file that is larger than memory (so that the reader cannot cache everything in memory and has to periodically go back to the file). Correctly handling this is non-trivial, and for now not on the immediate road-map.

**Comment by exonimbus** [comment]
I understand now your thinking, and you are correct of course. It is just that having concurrent access enabled in the fashion SQLite implements it, combined with all the immense analytic power packed inside DuckDB would make this a very hard to beat go to destination for an embedded database, in my humble opinion (think IOT for instance). I need this feature so badly that I am willing to go the route of creating a system visble global Event object in the writing process and hack the code furthermore to insure the scenarios you describe above will never happen in my particular case. I agree this is very hard to implement in a generic manner in DuckDB's case, and it is clearly not one of your priorities. In my use case, I could easily achieve this feature for a custom build solution, where writers and reader processes cooperate using a system level syncronization objects (think  a global session Windows Event object for instance). Btw, I noticed that even though my ugly hack kind of works most of the time, I have to always create a new DuckDB class instance and new Connection object in order to see the changes made from another process, otherwise the data is never updated for the reader process. This kind of sucks. I don't mind creating a new DuckDB object for each one of my reads but once the DB grows this is slow as hell. Is there any realistic way to disable internal caching and force reads straight from the .db file? I started to fiddle with checkpoint_wal_size and  use_direct_io = true but no luck so far. Like I said, I am hellbent to get this multiple readers one writer scenario work to the extent that I want it to. It is too big of a prize for me and my solution right now and I cannot give up when I feel that there must be a way to achieve this. I understand that you position your product as a strictly OLAP solution and you couldn't care less about OLTP scenarios. In my case I would love to put some kind of a network server on top of all this analytical power (i.e. mongoose web server for instance) and then make it into a low footprint/low cost computing node that one could report off of AND simultaneously log incoming process data. The possibilities would be truly unlimited that way, a dream come true. You could say that MariaDB does this already, but no, it doesn't. Plus it is a huge, awkward to install and not in-process solution, so no, thank you, I'll pass. I need something extra slim, statically linked with no dependencies. DuckDB fits the bill 100% in that regard.

**Comment by Alex-Monahan** [comment]
Could writing your new data out into csv or parquet files work for you? You could then just instantiate a new (empty or in memory) DuckDB instance on each read and scan those files. If you have an append only workflow, appending to a delimited file should be pretty fast. You can just have a separate file for each write process and then use DuckDB's glob syntax to query all of them. (You could also write a union all statement, but that isn't parallelized). You can specify the schema of the delimited file on read so you wouldn't have to diagnose the data types each time.

**Comment by exonimbus** [comment]
Of course writing to csv files and importing them periodically is a valid, legitimate workaround for this thorny issue. But if I go the csv route I don't need a database in the first place, so it kind of defeats the purpose of ever bothering using an of the shelve solution and I left building everything for scracth. My requirements are such that I need as real time as possible, I cannot import csv's in a loop in batches as I will miss certain transitions in the flux of data and cannot react on the incoming data in a next to real time fashion. That is the reason why people use databases in the first place, because they offer a consistent image as close to real time as possible of the current status of the system. Csv importing works for "eventually consistent" scenarios but not for "almost real-time consistent, sub-second time reference base"  scenarios.

**Comment by Mytherin** [comment]
> Btw, I noticed that even though my ugly hack kind of works most of the time, I have to always create a new DuckDB class instance and new Connection object in order to see the changes made from another process, otherwise the data is never updated for the reader process. This kind of sucks. I don't mind creating a new DuckDB object for each one of my reads but once the DB grows this is slow as hell. Is there any realistic way to disable internal caching and force reads straight from the .db file?

This is more or less what SQLite does, and what would be required to get multiple writers and multiple readers to work nicely together. The main focus on fixing that would be around optimizing the opening of a new database (which SQLite has optimized extensively, as you might imagine!).

Currently many things are being read on start-up that could be read lazily during execution (e.g. we could read block pages for columns as required in a query, rather than reading and setting everything up on startup).

> In my case I would love to put some kind of a network server on top of all this analytical power (i.e. mongoose web server for instance) and then make it into a low footprint/low cost computing node that one could report off of AND simultaneously log incoming process data. The possibilities would be truly unlimited that way, a dream come true. 

We do have a [REST server available for DuckDB](https://github.com/duckdb/duckdb/tree/master/tools/rest), although a full client/server database is out of scope, at least for now. 

> Of course writing to csv files and importing them periodically is a valid, legitimate workaround for this thorny issue. 

The solution Alex is proposing doesn't require you to import the files at all. You could create a view that includes a glob over a folder, e.g.:

```sql
CREATE VIEW my_view AS SELECT * FROM 'my_folder/*.csv';
```

And then query the view as if it were a table. The CSV files will be read every time the query is executed, and any new files added to the directory will be automatically processed by the system as well.

For CSV files this is rather slow (as all the text files need to be processed, etc), but for Parquet files this is not particularly problematic as the Parquet file format is columnar and rather fast to read. If you are running on many Parquet files it is recommended to use `PRAGMA enable_object_cache` to cache their metadata, though.

**Comment by dawningtruth** [comment]
As of Feb 2022, can DuckDB cater to the scenario of multiple reads and a single simultaneous writer.  

Or do you need to create some work around like a table with a switch that says, busy writing, come back later to read. Which is not ideal as it locks the database while updating.

**Comment by Alex-Monahan** [comment]
Hello!
The concurrency situation has not changed from the description above. It is also not currently on our roadmap to change this because it would be very difficult to support more concurrency while maintaining speed. The roadmap is determined by the DuckDB Foundation, so please let us know if you would like to contribute to it and become a member to help determine the roadmap!

**Comment by ededovic** [comment]
I would not mind your solution if read_only was working. But at the moment, it only works if all connections are read only, RO. Not only that, for update, all connections have to be closed. If I need to update the database, I have to find all processes that have connection to the database even the read only ones, and close them for that update to succeed, and then reopen those databases RO for additional analysis. It is a chore for one person team, but for multiple people team this makes duckdb not all that usable/sharable for analysis, the key feature. 

Few usage scenarios:
1. I analyze data in say DataGrip from JetBrains. If I forget to close that  read only connection, the next time a script that updates that database will fail. 
2. Similarly, if I have a Jupyter notebook open against the database, the RW script will fail. The worse issue in this case, I can not recall which one of the numerous Jupyter notebooks has the connection open. One alternative, is to copy a database if it is small, but for larger ones that is not an option.  
3. Working in a team, it is pretty hard to find who has the connection open even RO. 

It is easy to simulate failure by trying to open 2 database connection one RW and another one RO. Test done in duckdb==0.3.2.

> fn = 'lock_test.duckdb'
> drw = duckdb.connect(database=fn, read_only=False)
> dro = duckdb.connect(database=fn, read_only=True)


comes back with this error: 
> ---------------------------------------------------------------------------
> RuntimeError                              Traceback (most recent call last)
> <ipython-input-4-6d3108811e36> in <module>
>       1 fn = 'lock_test.duckdb'
>       2 drw = duckdb.connect(database=fn, read_only=False)
> ----> 3 dro = duckdb.connect(database=fn, read_only=True)
> 
> RuntimeError: IO Error: Cannot open file "lock_test.duckdb": The process cannot access the file because it is being used by another process.

If I flip that around, others say have RO and I just want to update something in the database with RW, same problem: 
> fn = 'lock_test.duckdb'
> dro = duckdb.connect(database=fn, read_only=True)
> drw = duckdb.connect(database=fn, read_only=False)

> RuntimeError                              Traceback (most recent call last)
> <ipython-input-8-aef332aab248> in <module>
>       1 fn = 'lock_test.duckdb'
>       2 dro = duckdb.connect(database=fn, read_only=True)
> ----> 3 drw = duckdb.connect(database=fn, read_only=False)
>       4 

> RuntimeError: IO Error: Cannot open file "lock_test.duckdb": The process cannot access the file because it is being used by another process.

**Comment by Alex-Monahan** [comment]
Hello and thank you for using DuckDB!

Unfortunately the behavior you are describing is expected. DuckDB can either support one read/write process, OR multiple read only processes. If you have a very transactional use case, we now have a SQLite reader so you could pull from SQLite while inserting or updating SQLite. We have the same for Postgres now also.

For analytical use cases, we often recommend in memory DuckDB instances that point to views built on top of parquet files. 

The REST server linked above is another option!

**Comment by arjenpdevries** [comment]
The suggestion by @Alex-Monahan is to read the SQLite database (where you make your updates) from DuckDB, for efficient analytical query processing. So the other direction, query SQLite from DuckDB instead of DuckDB from SQLite!

**Comment by Alex-Monahan** [comment]
Here is the sqlitescanner extension that you would need! I don't know if it is working on Windows just yet, but I haven't tested it in the last few weeks.
https://github.com/duckdblabs/sqlitescanner

**Comment by dawningtruth** [comment]
Thx Alex 😊.

Regards

Dave Lucas

[A person smiling for the camera  Description automatically generated with medium confidence]

Chief Executive Officer | Dawning Truth | Office: 066 233 1805 | Mobile: 082 407 3075 | dawningtruth.co.za<http://www.dawningtruth.co.za/>

[Text  Description automatically generated]

**Comment by inviridi** [comment]
I was so happy when I found DuckDB some days ago. I toyed around it it just worked so flawlessly out of the box. Thanks a lot for this wonderful experience and the fantastic, innovative DB you brought to the world!

My disappointment after reading this thread is more significant than I care to admit. I thought that multiple readers, one writer would be possible. Both, SQliteScanner and the Rest API look hacky to me, but what about shared access of the connection object in Python? I could implement one master process that serialises the DB requests. Performance is of secondary importance that important... 1 IoT writer, two or three readers are enough for my use case.

**Comment by Alex-Monahan** [comment]
In Python, each thread can make its own connection as long as they are in the same process. Having a single process handle DB requests should work!

**Comment by inviridi** [comment]
That's good news, @Alex-Monahan! I will pursue this route then. When you say "make its own connection", would you recommend setting up a connection in each thread like this:

```
import duckdb
con = duckdb.connect(database=':memory:')
```

... or is it recommended to have one connection object defined with the **global** keyword that is shared by all threads?

**Comment by hannes** [comment]
No, use `con.cursor()` or `con.duplicate()` to create a new connection for each thread, we

**Comment by Alex-Monahan** [comment]
I have also not tested this with the in memory connection string, but that may work! If it does not, you may need to refer to a physical DB file path so that all threads are connected to the same database.

**Comment by inviridi** [comment]
Thanks for the swift replies, @Alex-Monahan and @hannes! I slept over this issue and will probably approach it differently. I really like ZMQ and will fire up a separate process which listens to incoming ZMQ connections and hands the input to DuckDB. On a different port, the same process will wait for queries and return result dataframes. 

This can be done in a few lines of code and will work universally.

**Comment by Admolly** [comment]
@arjenpdevries does this have any noticeable impact  on query performance?  Definitely interested in implementing this if not.
